### PR TITLE
Enable access token cache.

### LINKED ISSUE #23473: Azure Keyvault JCA gets access token remotely every time it visits key vault.
Azure keyvault jca needs to visit key vault remotely. For each visit, it also needs an access token, so currently it also gets the access token remotely each time. This is not what we want, because the same access token can be reused, and to get the remote access token is time costly.

Please see the log:
 c.a.security.keyvault.jca.AuthClient     : Getting access token using client ID / client secret
c.a.s.keyvault.jca.KeyVaultClient        : Getting key for alias: friend
c.a.security.keyvault.jca.AuthClient     : Getting access token using client ID / client secret
c.a.security.keyvault.jca.AuthClient     : Getting access token using client ID / client secret
c.a.s.keyvault.jca.KeyVaultClient        : Getting certificate for alias: friend
c.a.security.keyvault.jca.AuthClient     : Getting access token using client ID / client secret
c.a.s.keyvault.jca.KeyVaultClient        : Using Azure Key Vault: https://kv-sp.vault.azure.net/
c.a.security.keyvault.jca.AuthClient     : Getting access token using client ID / client secret
c.a.s.keyvault.jca.KeyVaultClient        : Getting key for alias: friend
c.a.security.keyvault.jca.AuthClient     : Getting access token using client ID / client secret
c.a.security.keyvault.jca.AuthClient     : Getting access token using client ID / client secret
c.a.s.keyvault.jca.KeyVaultClient        : Getting certificate for alias: friend
c.a.security.keyvault.jca.AuthClient     : Getting access token using client ID / client secret
o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8443 (https) with context path ''
c.a.s.b.s.w.SpringBootWebappApplication  : Started SpringBootWebappApplication in 117.996 seconds (JVM running for 118.846)

### PR DESCRIPTION
**Pull request by michaelqi793**
Fixes #23473

### CODE REVIEW

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/model/OAuthToken.java [hunk 1]
```diff
@@ -4,6 +4,8 @@
 
 import com.fasterxml.jackson.annotation.JsonProperty;
 
+import java.time.OffsetDateTime;
+
 /**
  * An OAuth2 token.
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/model/OAuthToken.java:10]:
```
public class AccessToken {
```

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java [hunk 2]
```diff
@@ -74,6 +74,25 @@ public static String getAccessToken(String resource, String identity) {
         return result;
     }
 
+    /**
+     * Get an access token for a managed identity.
+     *
+     * @param resource the resource.
+     * @param identity the user-assigned identity (null if system-assigned)
+     * @return the authorization token.
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java:82]: @return the access token.

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java [hunk 3]
```diff
@@ -121,14 +157,25 @@ public static String getAccessToken(String resource, String aadAuthenticationUrl
...
+    }
+
+    /**
+     * Get the access token on Azure App Service.
+     *
+     * @param resource the resource.
+     * @param clientId the user-assigned managed identity (null if system-assigned).
+     * @return the authorization token.
+     */
+    private static OAuthToken getAccTokenOnAppService(String resource, String clientId) {
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java:172]: `client-id` -> `managedIdentity`

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java [hunk 4]
```diff
@@ -121,14 +157,25 @@ public static String getAccessToken(String resource, String aadAuthenticationUrl
...
         LOGGER.entering("AccessTokenUtil", "getAccessTokenOnAppService", resource);
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java:173]: the log's information is not same to this method name. I suggest change the method name.

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java (hunk 3 above)
Comment (chenrujun) [reply sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java:172]: Can we remove all methods that return `String`? just return `OAuthToken`.

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java [hunk 5]
```diff
@@ -74,6 +74,25 @@ public static String getAccessToken(String resource, String identity) {
...
+     *
+     * @param resource the resource.
+     * @param identity the user-assigned identity (null if system-assigned)
+     * @return the authorization token.
+     */
+    public static OAuthToken getAccToken(String resource, String identity) {
+        OAuthToken result;
+
+        if (System.getenv("WEBSITE_SITE_NAME") != null
+            && !System.getenv("WEBSITE_SITE_NAME").isEmpty()) {
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java:88]: Can we remove all methods that return `String`? just return `AccessToken`.

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java [hunk 6]
```diff
@@ -94,23 +130,23 @@ public static String getAccessToken(String resource, String aadAuthenticationUrl
 
         StringBuilder oauth2Url = new StringBuilder();
         oauth2Url.append(aadAuthenticationUrl == null ? OAUTH2_TOKEN_BASE_URL : aadAuthenticationUrl)
-                 .append(tenantId)
-                 .append(OAUTH2_TOKEN_POSTFIX);
+            .append(tenantId)
+            .append(OAUTH2_TOKEN_POSTFIX);
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java:134]: Please don't bring unnecessary change.

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java [hunk 7]
```diff
@@ -94,23 +130,23 @@ public static String getAccessToken(String resource, String aadAuthenticationUrl
...
 
         StringBuilder requestBody = new StringBuilder();
         requestBody.append(GRANT_TYPE_FRAGMENT)
-                   .append(CLIENT_ID_FRAGMENT).append(clientId)
-                   .append(CLIENT_SECRET_FRAGMENT).append(clientSecret)
-                   .append(RESOURCE_FRAGMENT).append(resource);
+            .append(CLIENT_ID_FRAGMENT).append(clientId)
+            .append(CLIENT_SECRET_FRAGMENT).append(clientSecret)
+            .append(RESOURCE_FRAGMENT).append(resource);
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java:140]: Same here.

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java [hunk 8]
```diff
@@ -121,14 +157,25 @@ public static String getAccessToken(String resource, String aadAuthenticationUrl
...
         LOGGER.entering("AccessTokenUtil", "getAccessTokenOnAppService", resource);
         LOGGER.info("Getting access token using managed identity based on MSI_SECRET");
-        String result = null;
-
         StringBuilder url = new StringBuilder();
         url.append(System.getenv("MSI_ENDPOINT"))
-           .append("?api-version=2017-09-01")
-           .append(RESOURCE_FRAGMENT).append(resource);
+            .append("?api-version=2017-09-01")
+            .append(RESOURCE_FRAGMENT).append(resource);
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java:178]: Don't bring unnecessary change.

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java [hunk 9]
```diff
@@ -141,10 +188,10 @@ private static String getAccessTokenOnAppService(String resource, String clientI
 
         if (body != null) {
             OAuthToken token = (OAuthToken) JsonConverterUtil.fromJson(body, OAuthToken.class);
-            result = token.getAccessToken();
+            LOGGER.exiting("AccessTokenUtil", "getAccessTokenOnAppService");
+            return token;
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java:192]: Don't bring unnecessary change.

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java [hunk 10]
```diff
@@ -155,16 +202,28 @@ private static String getAccessTokenOnAppService(String resource, String clientI
...
         LOGGER.info("Getting access token using managed identity");
         if (identity != null) {
             LOGGER.log(INFO, "Using managed identity with object ID: {0}", identity);
         }
-        String result = null;
 
         StringBuilder url = new StringBuilder();
         url.append(OAUTH2_MANAGED_IDENTITY_TOKEN_URL)
-           .append(RESOURCE_FRAGMENT).append(resource);
+            .append(RESOURCE_FRAGMENT).append(resource);
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java:226]: Same here.

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java [hunk 11]
```diff
@@ -175,9 +234,9 @@ private static String getAccessTokenOnOthers(String resource, String identity) {
 
         if (body != null) {
             OAuthToken token = (OAuthToken) JsonConverterUtil.fromJson(body, OAuthToken.class);
-            result = token.getAccessToken();
+            LOGGER.exiting("AccessTokenUtil", "getAccessTokenOnOthers");
+            return token;
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java:238]: Same here.

**Comment by chenrujun** [comment]
Hi, @michaelqi793 , if this PR is ready to ready for review, please click `Ready for review` button.

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/model/OAuthToken.java [hunk 12]
```diff
@@ -15,6 +17,35 @@
...
+    /**
+     * Stores the life duration of the access token.
+     */
+    @JsonProperty("expires_in")
+    private long expiresIn;
+
+    /**
+     * Stores the life duration of the access token.
+     */
+    @JsonProperty("ext_expires_in")
```
[the remark this answers is not in the document]
Comment (michaelqi793) [reply sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/model/OAuthToken.java:29]: Yes, and they have the same value, currently.

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java (hunk 5 above)
Comment (michaelqi793) [reply sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java:88]: I think Yes. But there is a public method that returns string. Not sure whether it is used outside our sight.

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/KeyVaultClient.java [hunk 13]
```diff
@@ -183,8 +189,21 @@ public static KeyVaultClient createKeyVaultClientBySystemProperty() {
      * @return the access token.
      */
     private String getAccessToken() {
+        if (cacheToken != null && !cacheToken.isExpired()) {
+            return cacheToken.getAccessToken();
+        }
+        cacheToken = getAccessTokenByHttpRequest();
```
Comment (backwind1233) [inline sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/KeyVaultClient.java:195]: do we need a lock here?

Comment (michaelqi793) [reply sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/KeyVaultClient.java:195]: > do we need a lock here?

Hi, KeyVaultClient is not designed for thread-safety use and there is no concurrent use case for it.

**Comment by vcolin7** [comment]
I haven't looked at the implementation yet, but here's the class we use from Azure Core that handles this type of authentication: [BearerTokenAuthenticationPolicy](https://github.com/Azure/azure-sdk-for-java/blob/334793a7fc421b7eb1f0b495ed2c67ef4813fe6b/sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java). Hope it helps for something :P

**Comment by vcolin7** [comment]
I haven't looked at the implementation yet, but here's the class we use from Azure Core that handles this type of authentication: [BearerTokenAuthenticationPolicy](https://github.com/Azure/azure-sdk-for-java/blob/334793a7fc421b7eb1f0b495ed2c67ef4813fe6b/sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java). Hope it helps for something :P

**Comment by michaelqi793** [comment]
> I haven't looked at the implementation yet, but here's the class we use from Azure Core that handles this type of authentication: [BearerTokenAuthenticationPolicy](https://github.com/Azure/azure-sdk-for-java/blob/334793a7fc421b7eb1f0b495ed2c67ef4813fe6b/sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java). Hope it helps for something :P

Thanks Vcolin! key vault jca should be light, it does not use this chain of responsibility currently.

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/KeyVaultClient.java [hunk 14]
```diff
@@ -116,6 +117,11 @@ public static String getAADLoginURIByKeyVaultBaseUri(String keyVaultBaseUri) {
      */
     private String managedIdentity;
 
+    /**
+     * Stores the token.
+     */
+    private AccessToken cacheToken;
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/KeyVaultClient.java:123]: nit: rename to accessToken.

**Review by chenrujun — APPROVED** [review]
LGTM.
